src/Language/Bitcoin/Script/Descriptors/Utils.hs

```diff
@@ -140,6 +165,29 @@ compile = \case
   where
     compileMaybe = eitherToMaybe . M.compile
 
+{- | Produce the MAST described by the tree descriptor. Fails when any keys in
+ the descriptor are indeterminate.
+-}
+compileTree :: TreeDescriptor -> Maybe MAST
+compileTree = \case
+    TapLeaf script -> MASTLeaf 0xc0 <$> compileTapLeaf script
+    TapBranch left right ->
+        MASTBranch <$> compileTree left <*> compileTree right
+
+{- | Produce the script described by the descriptor in a taproot leaf context.
+Fails when any keys in the descriptor are indeterminate.
+-}
+compileTapLeaf :: ScriptDescriptor -> Maybe Script
+compileTapLeaf = \case
+    Pk keyDesc -> do
+        pubKey <- keyDescPubKey keyDesc
+        return $
+            Script
+                [ opPushData $ encode $ XOnlyPubKey $ pubKeyPoint pubKey
+                , OP_CHECKSIG
+                ]
```

Comment:
This can and should support all of the branches of `ScriptDescriptor`.  That means it can also fail if `Raw` includes an invalid script.

Reply:
My understanding of the spec is that the only script expression of `ScriptDescriptor` that is currently valid as a taproot leaf is `pk()`. `pkh()`, `multi()` and `sortedmulti()` are all not allowed in a `tr()` expression tree for now. I think there will be new `multi_a()`  and `sortedmulti_a()` multisig expressions that will compile to the new `OP_CHECKSIGADD` opcode added later on specifically for taproot. `raw()` also cannot occur here as it is supposed to be a top-level expression: https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md#reference.

I don't think there's anything illegal about having these scripts in a TR output, just not for descriptors. I see the function as only compiling taproot leaf scripts that are allowed within a taproot descriptor, not a general `ScriptDescriptor -> Script` function so maybe this shouldn't be exported or a separate `TapScriptDescriptor` should be introduced?

Reply:
Again, misread the spec.  Also it is too bad that `Raw` is a constructor of `ScriptDescriptor`.  I can't remember if the restiction that it must be top-level predates the library or not.  Anyway, I agree with your reading.  Would you add a comment making it clear why the other branches lead to nothing?

Reply:
Sure, I've expanded the comments. Let me know if you want me to add more detail.